src/System.Text.Json/tests/JsonObjectTests.cs

```diff
@@ -0,0 +1,609 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using Xunit;
+
+namespace System.Text.Json.Tests
+{
+    public static partial class JsonObjectTests
+    {
+        [Fact]
+        public static void TestDefaultConstructor()
+        {
+            var jsonObject = new JsonObject();
+            Assert.Equal(0, jsonObject.PropertyNames.Count);
+            Assert.Equal(0, jsonObject.Values.Count);
+        }
+
+        [Fact]
+        public static void TestIEnumerableKVPConstructor()
+        {
+            var jsonProperties = new List<KeyValuePair<string, JsonNode>>();
+            jsonProperties.Add(new KeyValuePair<string, JsonNode>("number", new JsonNumber(17)));
+            jsonProperties.Add(new KeyValuePair<string, JsonNode>("string", new JsonString("property value")));
+            jsonProperties.Add(new KeyValuePair<string, JsonNode>("boolean", new JsonBoolean(true)));
+
+            var jsonObject = new JsonObject(jsonProperties);
+
+            Assert.IsType<JsonNumber>(jsonObject["number"]);
+            Assert.Equal(17, (jsonObject["number"] as JsonNumber).GetInt32());
+
+            Assert.IsType<JsonString>(jsonObject["string"]);
+            Assert.Equal("property value", (jsonObject["string"] as JsonString).Value);
+
+            Assert.IsType<JsonBoolean>(jsonObject["boolean"]);
+            Assert.True((jsonObject["boolean"] as JsonBoolean).Value);
+        }
+
+        private enum ExpectedValue
+        {
+            Previous,
+            New
+        }
+
+        private static void TestDuplicates(DuplicatePropertyNameHandling duplicatePropertyNameHandling, ExpectedValue expected = default, bool valueConstructor = true)
+        {
+            string previousValue = "value1";
+            string newValue = "value1";
+
+            JsonObject jsonObject = valueConstructor ? new JsonObject(duplicatePropertyNameHandling) : new JsonObject();
+            jsonObject.Add("property", new JsonString(previousValue));
+
+            Assert.IsType<JsonString>(jsonObject["property"]);
+            Assert.Equal(previousValue, (jsonObject["property"] as JsonString).Value);
+
+            jsonObject.Add("property", new JsonString(newValue));
+            Assert.IsType<JsonString>(jsonObject["property"]);
+
+            var expectedString = expected == ExpectedValue.Previous ? previousValue : newValue;
+
+            Assert.Equal(expectedString, (jsonObject["property"] as JsonString).Value);
+        }
+
+        [Fact]
+        public static void TestDuplicatesReplace()
```

Comment:
These tests can be merged into a Theory where you pass in the DuplicatePropertyNameHandling enum as InlineData.